### PR TITLE
added hook for user address update

### LINKED ISSUE #4427: Update profile from My Account gives a 500 error on /api/user/me on validation fail and user is not aware of save failure
## Current behavior
Update profile from `My Account` gives a 500 error on `/api/user/me` on validation fail and user is not:
- aware of save fail
- aware of fields that failed validation
- able to change data that failed validation.

## Expected behavior
Update profile from `My Account` on validation fail, user is:
- aware of save fail
- aware of fields that failed validation
- able to change data that failed validation.

## Steps to reproduce the issue
On a project, users are unable to update their profile from `/my-account`.
I am able to trigger a similar behavior on https://demo.vuestorefront.io.

I create an account on https://demo.vuestorefront.io.
I go to https://demo.vuestorefront.io/my-account/shipping-details.
I fill all fields except "Phone Number"

demo vuestorefront io_my-account_shipping-details

I click "Update my shipping details"
Filled field appear on https://demo.vuestorefront.io/my-account/shipping-details

demo vuestorefront io_my-account_shipping-details (1)

I reload https://demo.vuestorefront.io/my-account/shipping-details page and then I become aware that something went wrong as my shipping details are lost. 

demo vuestorefront io_my-account_shipping-details (2)

There is no error message in html.
User has no way to know that an error occurred ans will not be able to change data that failed validation.

In browser developer tools Network tab, I see a 500 error on: https://demo.storefrontcloud.io/api/user/me

```
{"code":500,"result":{"errorMessage":"\"telephone\" is required. Enter and try again.","code":400}}
```

I have similar errors on a project :

```
{"code":500,"result":[{"keyword":"maxItems","dataPath":".customer.addresses","schemaPath":"#/properties/customer/properties/addresses/maxItems","params":{"limit":2},"message":"should NOT have more than 2 items"}]}
```

```
{"code":500,"result":[{"keyword":"minLength","dataPath":".customer.addresses[0].postcode","schemaPath":"#/properties/customer/properties/addresses/items/properties/postcode/minLength","params":{"limit":5},"message":"should NOT be shorter than 5 characters"}]}
```

## Can you handle fixing this bug by yourself?

- [ ] YES
- [x] NO

## Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer.

- [ ] This is a bug report for test version on https://test.storefrontcloud.io - In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
- [ ] This is a bug report for current Release Candidate version on https://next.storefrontcloud.io - In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
- [x] This is a bug report for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version hotfix - In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

## Environment details

- Browser: 
- OS: 
- Node: 
- Code Version:

### PR DESCRIPTION
**Pull request by gibkigonzo**
### Related Issues

closes #4427

### Short Description and Why It's Useful

Added 'afterUserProfileUpdated' hook which is triggered after profile is updated (saved in vuex)
Added 'beforeUserProfileUpdate' mutator which can be used to extend user profile before save.
Changes to user address update flow. Now we can take control over address update in components. We can handle errors and edit mode will be closed after task is finished.
Also fixed bug with `config.users.allowModification` - we need to be sure that we send properties that are specified in `config.users.allowModification`

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [x] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [x] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: core/modules/user/components/UserShippingDetails.ts [hunk 1]
```diff
@@ -111,18 +112,29 @@ export const UserShippingDetails = {
...
       this.$bus.$emit('myAccount-before-updateUser', updatedShippingDetails)
-      if (!updatedShippingDetails) {
-        this.shippingDetails = this.getShippingDetails()
-        this.useCompanyAddress = false
-        this.remainInEditMode = false
-      }
-      if (!this.remainInEditMode) {
-        this.isEdited = false
-      }
+      userHooks.afterUserProfileUpdated(async (event: any) => {
```
Comment (Fifciu) [inline core/modules/user/components/UserShippingDetails.ts:120]: Isn't it type `Task` from `core/lib/sync/types/Task.ts`?

File: core/modules/user/components/UserAccount.ts [hunk 2]
```diff
@@ -150,25 +151,36 @@ export const UserAccount = {
...
-      if (!updatedProfile) {
-        this.currentUser = Object.assign({}, this.$store.state.user.current)
-        this.userCompany = this.getUserCompany()
-        this.changePassword = false
-        this.oldPassword = ''
-        this.password = ''
-        this.rPassword = ''
-        if (!this.userCompany.company) {
-          this.addCompany = false
+      userHooks.afterUserProfileUpdated(async (event) => {
```
Comment (Fifciu) [inline core/modules/user/components/UserAccount.ts:159]: Check comment below at first please.
If it is type Task - let's append it there
if not let's remove parentheses

File: core/modules/user/components/UserAccount.ts [hunk 3]
```diff
@@ -150,25 +151,36 @@ export const UserAccount = {
...
+            if (!this.userCompany.company) {
+              this.addCompany = false
+            }
+            this.remainInEditMode = false
+          }
+          if (!this.remainInEditMode) {
+            this.isEdited = false
+          }
+        } else {
+          await this.$store.dispatch('notification/spawnNotification', {
```
Comment (Fifciu) [inline core/modules/user/components/UserAccount.ts:177]: Do we need `await`? I see no Promise in spawnNotification:
```ts
spawnNotification ({ commit, state, dispatch }, notification: NotificationItem): NotificationItem {
      if (state.notifications.length > 0 &&
        state.notifications[state.notifications.length - 1].message === notification.message
      ) {
        return
      }

      const id = Math.floor(Math.random() * 100000)
      const newNotification = { id, ...notification }

      commit('add', newNotification)

      if (!newNotification.hasNoTimeout) {
        setTimeout(() => {
          dispatch('removeNotificationById', id)
        }, newNotification.timeToLive || 5000)
      }

      return newNotification
    }
```

File: core/modules/user/components/UserShippingDetails.ts [hunk 4]
```diff
@@ -111,18 +112,29 @@ export const UserShippingDetails = {
...
+          if (!updatedShippingDetails) {
+            this.shippingDetails = this.getShippingDetails()
+            this.useCompanyAddress = false
+            this.remainInEditMode = false
+          }
+          if (!this.remainInEditMode) {
+            this.isEdited = false
+          }
+        } else {
+          await this.$store.dispatch('notification/spawnNotification', {
```
Comment (Fifciu) [inline core/modules/user/components/UserShippingDetails.ts:131]: Same question as above with this `await`

File: core/modules/user/store/actions.ts [hunk 5]
```diff
@@ -167,6 +167,7 @@ const actions: ActionTree<UserState, RootState> = {
    * Update user profile with data from My Account page
    */
   async update (_, profile: UserProfile) {
+    profile = userHooksExecutors.beforeUserProfileUpdate(profile)
     await UserService.updateProfile(profile, 'user/handleUpdateProfile')
   },
   async handleUpdateProfile ({ dispatch }, event) {
```
Comment (Fifciu) [inline core/modules/user/store/actions.ts:173]: type Task?

File: core/modules/user/store/actions.ts [hunk 6]
```diff
@@ -167,6 +167,7 @@ const actions: ActionTree<UserState, RootState> = {
    * Update user profile with data from My Account page
    */
   async update (_, profile: UserProfile) {
+    profile = userHooksExecutors.beforeUserProfileUpdate(profile)
```
Comment (Fifciu) [inline core/modules/user/store/actions.ts:170]: I wonder (I am not sure) - maybe should we send copy of `profile` to this function? Without that, `profile` would be mutable inside the `beforeUserProfileUpdate` - which could cause hard to debug side effects. Wdyt?

**Review by Fifciu — CHANGES_REQUESTED** [review]

File: core/modules/user/components/UserAccount.ts (hunk 2 above)
Comment (gibkigonzo) [reply core/modules/user/components/UserAccount.ts]: it is type Task but we don't need to add it here. It's already added as generic type https://github.com/DivanteLtd/vue-storefront/pull/4929/files#diff-2cde0f6fb20851a4e391587c17b3dad1R22
parentheses removed :+1:

File: core/modules/user/components/UserAccount.ts (hunk 3 above)
Comment (gibkigonzo) [reply core/modules/user/components/UserAccount.ts]: nope, removed

File: core/modules/user/components/UserShippingDetails.ts (hunk 1 above)
Comment (gibkigonzo) [reply core/modules/user/components/UserShippingDetails.ts]: same as above. It already has type so I shouldn't add here any.

File: core/modules/user/components/UserShippingDetails.ts (hunk 4 above)
Comment (gibkigonzo) [reply core/modules/user/components/UserShippingDetails.ts]: done

File: core/modules/user/store/actions.ts (hunk 5 above)
Comment (gibkigonzo) [reply core/modules/user/store/actions.ts]: done

File: core/modules/user/store/actions.ts (hunk 6 above)
Comment (gibkigonzo) [reply core/modules/user/store/actions.ts:171]: I think it's more a question about how mutator should work. Right now mutator mutate object. Maybe mutator should then always return a copy? You can check that it used in that way in whole app :thinking: Everywhere its used always once, so the question is what will happen if we will use it more then once. Probably there will be random order of mutations

Comment (Fifciu) [reply core/modules/user/store/actions.ts:171]: Not good. We have to discuss it internally

**Review by Fifciu — APPROVED** [review]

**Pull request merged**